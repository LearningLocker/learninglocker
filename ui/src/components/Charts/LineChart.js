import React from 'react';
import { AutoSizer } from 'react-virtualized';
import { LineChart as Chart, XAxis, YAxis, Line, CartesianGrid } from 'recharts';
import NoData from 'ui/components/Graphs/NoData';
import { compose } from 'recompose';
import { wrapLabel } from 'ui/utils/defaultTitles';
import {
  getResultsData,
  getShortModel,
  getChartData,
  hasData,
  renderTooltips,
  renderLegend,
  hiddenSeriesState
} from './Chart';
import {
  Chart as StyledChart,
  BarContainer,
  XAxis as StyledXAxis,
  YAxis as StyledYAxis,
  XAxisLabel,
  ChartWrapper
} from './styled';

const sortData = data =>
  data.sortBy(e => e.get('id'));

const getSortedData = results => labels =>
  sortData(getResultsData(results)(labels));

const renderLine = colors => (label, i) => (
  <Line
    key={i + 1}
    dataKey={`Series ${i + 1}`}
    type="monotone"
    dot={false}
    stroke={colors.get(i)}
    name={label} />
);

const renderLines = labels => colors =>
  labels.map(renderLine(colors)).valueSeq();

const renderLineChart = (labels, toggleHiddenSeries, hiddenSeries) => colors => data => ({ width, height }) => (
  <Chart
    data={getChartData(data, hiddenSeries)}
    width={width}
    height={height}
    margin={{ top: 10, right: 35, left: -20, bottom: 5 }}>
    <XAxis type="category" dataKey="cellId" tickFormatter={getShortModel(data)} />
    <YAxis type="number" />
    <CartesianGrid strokeDasharray="1 1" />
    {renderLegend(labels, toggleHiddenSeries)}
    {renderLines(labels)(colors)}
    {renderTooltips(data)}
  </Chart>
);

const renderChart = (component, axesLabels, chartWrapperFn, model) => (
  <StyledChart>
    <BarContainer>
      <StyledYAxis>
        {wrapLabel(axesLabels.yLabel || model.getIn(['axesvalue', 'searchString'], 'Y-Axis'))}
      </StyledYAxis>
      <ChartWrapper>
        {chartWrapperFn(component)}
      </ChartWrapper>
    </BarContainer>
    <XAxisLabel>
      <StyledXAxis>
        {axesLabels.xLabel || 'yyyy/mm/dd'}
      </StyledXAxis>
    </XAxisLabel>
  </StyledChart>
);

const renderChartResults = (labels, toggleHiddenSeries, hiddenSeries) => colors => results =>
  renderLineChart(labels, toggleHiddenSeries, hiddenSeries)(colors)(getSortedData(results)(labels));

const renderResults = results => (labels, toggleHiddenSeries, hiddenSeries) => colors => axesLabels => chartWrapperFn => model =>
  renderChart(renderChartResults(labels, toggleHiddenSeries, hiddenSeries)(colors)(results), axesLabels, chartWrapperFn, model);

export default compose(hiddenSeriesState)((
  {
    results,
    labels,
    colors,
    axesLabels,
    chartWrapperFn = component => (<AutoSizer>{component}</AutoSizer>),
    toggleHiddenSeries,
    hiddenSeries,
    model
  }) =>
    (
      hasData(results)
        ? renderResults(results)(labels, toggleHiddenSeries, hiddenSeries)(colors)(axesLabels)(chartWrapperFn)(model)
        : <NoData />
    )
);
