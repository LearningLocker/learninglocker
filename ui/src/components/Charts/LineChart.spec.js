import React from 'react';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import { fromJS } from 'immutable';
import LineChart from './LineChart';

test('LineChartResult should render axes', () => {
  const mockResult = fromJS([[{
    '1970-01-01': {
      _id: '1970-01-01',
      count: 1,
      model: '1970-01-01'
    }
  }]]);

  const lineChart = renderer
    .create(
      <LineChart
        results={mockResult}
        labels={fromJS([undefined])}
        axesLabels={{
          yLabel: 'THE-Y',
          xLabel: 'THE-X'
        }}
        chartWrapperFn={() => (<div />)} />
    )
    .toJSON();

  expect(lineChart).toMatchSnapshot();
});
