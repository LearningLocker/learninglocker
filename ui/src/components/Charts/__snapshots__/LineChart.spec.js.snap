// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChartResult should render axes 1`] = `
.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 svg g text {
  font-size: 12px;
}

.c3 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c5 {
  display: inline-block;
  white-space: nowrap;
}

.c4 {
  text-align: right;
  word-break: break-word;
  overflow: hidden;
  text-overflow: ellipsis;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: static;
  margin-right: 15px;
  margin-bottom: 10px;
}

.c2 {
  display: inline-block;
  -webkit-writing-mode: vertical-rl;
  -ms-writing-mode: tb-rl;
  writing-mode: vertical-rl;
  -webkit-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
  margin: auto 0;
  white-space: pre-wrap;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      THE-Y
    </div>
    <div
      className="c3"
    >
      <div />
    </div>
  </div>
  <div
    className="c4"
  >
    <span
      className="c5"
    >
      THE-X
    </span>
  </div>
</div>
`;
